lucet-runtime/lucet-runtime-internals/src/region/mmap.rs

```diff
@@ -80,38 +80,42 @@ impl RegionInternal for MmapRegion {
         module: Arc<dyn Module>,
         embed_ctx: CtxMap,
         heap_memory_size_limit: usize,
+        alloc_strategy: AllocStrategy,
     ) -> Result<InstanceHandle, Error> {
-        let custom_limits;
-        let mut limits = self.get_limits();
-
-        // Affirm that the module, if instantiated, would not violate
-        // any runtime memory limits.
-        match heap_memory_size_limit.cmp(&limits.heap_memory_size) {
-            Ordering::Less => {
-                // The supplied heap_memory_size is smaller than
-                // default. Augment the limits with this custom value
-                // so that it may be validated.
-                custom_limits = Limits {
-                    heap_memory_size: heap_memory_size_limit,
-                    ..*limits
-                };
-                limits = &custom_limits;
+        let limits = self.get_limits();
+        module.validate_runtime_spec(&limits, heap_memory_size_limit)?;
+
+        let slot;
+
+        match alloc_strategy {
+            AllocStrategy::Linear => {
+                slot = self
+                    .freelist
+                    .write()
+                    .unwrap()
+                    .pop()
+                    .ok_or(Error::RegionFull(self.capacity))?
             }
-            Ordering::Equal => (),
-            Ordering::Greater => {
-                return Err(Error::InvalidArgument(
-                    "heap memory size requested for instance is larger than slot allows",
-                ))
+            AllocStrategy::Random => {
+                let mut free_slot_vector = self.freelist.write().unwrap();
+                if free_slot_vector.len() == 0 {
+                    return Err(Error::RegionFull(self.capacity));
+                }
+                // Instantiate a random number generator and get a random slot.
+                let mut rng = thread_rng();
+                let rnd_idx = rng.gen_range(0, free_slot_vector.len());
+                slot = free_slot_vector.swap_remove(rnd_idx);
+            }
+            AllocStrategy::CustomRandom(mut rng) => {
```

Comment:
It feels a little bit weird to expose something only intended to be used internally in the public API.

Do the tests really need `CustomRandom`? Having non-deterministic input may actually be better to trigger edge cases that a deterministic sequence would never trigger, but real-world usage would.

Reply:
I see your point.  So much so that I began writing the tests non-deterministically, as you see here:  https://github.com/bytecodealliance/lucet/pull/496/commits/421578f2730a7b9c4082bc8f02e912eee21c5330.  😊  I agree that non-deterministic input may trigger issues that could happen in real-world use.  

In talking with @acfoltzer about crafting these tests, he expressed a preference for them to be deterministic.  This approach has its own advantage.  He expressed a worry about having tests in our repo that could break 1 time out of 100, which is a difficult error situation to debug.

Acknowledging the many times that I have endured flakey tests in other repos I have worked on, my own preference is to have non-deterministic tests for this scenario.  But it's only a slight preference and I acknowledge that I do not know this repo as well as others on the team.  It would be great to hear more conversation on this matter.

Reply:
In wiggle (wasmtime repo) we use `proptest`s in CI. When those tests fail, they print out the seed and instructions to use that seed to reproduce locally. Can we use that approach, either re-creating it or better yet just using proptest's machinery?

Reply:
It should be possible to do the same here; use `thread_rng()` to create a random seed, print it, and then seed the `StdGen` that the test uses. I really do not want to have to rely on getting lucky in order to reproduce breakages.

Reply:
Thanks for the discussion.  Here's what I'm hearing:
- It is undesirable to rely on lucky seeds.
- It is undesirable to rely on luck to reproduce breaks.
- Let's keep the `CustomRandom` enum in the `AllocStrategy`.
- Use `thread_rng()` to create a random seed and print it so that any breaking seeds are immediately known at the time of failure.

Do I have that right?